wgsl/index.bs

```diff
@@ -931,9 +941,230 @@ a result value, such as the non-result-value effects of its subexpressions.
 
 TODO: example: non-result-value effect is any side effect of a function call subexpression.
 
+## Ideal Types ## {#ideal-types-sec}
+
+There are two <dfn noexport>ideal types</dfn>:
+* The <dfn noexport>IdealInteger</dfn> type is the set of all integers.
+    * An [=syntax/int_literal=] without an `i` or `u` suffix denotes an [=IdealInteger=] value.
+* The <dfn noexport>IdealReal</dfn> type is the set of all real numbers (e.g. 1.25, -100, &pi;)
+    * A [=syntax/float_literal=] without an `f` suffix denotes a [=IdealReal=] value.
+
+A type is <dfn noexport>concrete</dfn> if it is not [=ideal type|ideal=].
+
+A [=numeric literal=] without a suffix denotes a value in an ideal type.
+However, each numeric value in memory, in pipeline I/O, or used in most
+computations belongs to one of the [=numeric scalar=] types.
+
+Converting a value from an ideal type to a numeric scalar type is called <dfn noexport>materialization</dfn>.
+* An [=IdealReal=] value can only convert to [=f32=].
+* An [=IdealInteger=] value can only convert to one of the [=numeric scalar=] types.
```

Comment:
I wonder whether it is cleaner to not allow ideal integers to convert to f32. Instead, rely on the promotion rule below to first promote to ideal real.
```suggestion
* An [=IdealInteger=] value can only convert to [=i32=], [=u32]= or be [=promotion rule|promoted=] to [=IdealReal=].
```